pkg/networkutils/network.go

```diff
@@ -549,3 +629,134 @@ func isNetworkUnreachable(err error) bool {
 	}
 	return false
 }
+
+// GetRuleList returns IP rules
+func (n *linuxNetwork) GetRuleList() ([]netlink.Rule, error) {
+	return n.netLink.RuleList(unix.AF_INET)
+}
+
+// GetRuleListBySrc returns IP rules whose has matching source IP
+func (n *linuxNetwork) GetRuleListBySrc(ruleList []netlink.Rule, src net.IPNet) ([]netlink.Rule, error) {
+	var srcRuleList []netlink.Rule
+
+	for _, rule := range ruleList {
+		if rule.Src != nil && rule.Src.IP.Equal(src.IP) {
+			srcRuleList = append(srcRuleList, rule)
+		}
+	}
+
+	return srcRuleList, nil
+}
+
+// DelRuleListBySrc deletes IP rules who has matcing source IP
+func (n *linuxNetwork) DelRuleListBySrc(src net.IPNet) error {
+
+	log.Infof("Delete Rule List By Src [%v", src)
+
+	ruleList, err := n.GetRuleList()
+	if err != nil {
+		log.Errorf("DelRuleListBySrc: failed to get rule list %v", err)
+		return err
+	}
+
+	srcRuleList, err := n.GetRuleListBySrc(ruleList, src)
+
+	if err != nil {
+		log.Errorf("DelRuleListBySrc: failed to retrieve rule list %v", err)
+		return err
+	}
+
+	log.Infof("Remove current list [%v]", srcRuleList)
+
+	for _, rule := range srcRuleList {
+		if n.netLink.RuleDel(&rule); err != nil && !containsNoSuchRule(err) {
+			log.Errorf("Failed to cleanup old IP rule: %v", err)
+			return errors.Wrapf(err, "DelRuleListBySrc: failed to delete old rule")
+		}
+
+		var toDst string
+
+		if rule.Dst != nil {
+			toDst = rule.Dst.String()
+		}
+		log.Debugf("DelRuleListBySrc: Successfully removed current rule [%v] to %s", rule, toDst)
+	}
+
+	return nil
+}
+
+// UpdateRuleListBySrc modify IP rules who has matching IP source
+func (n *linuxNetwork) UpdateRuleListBySrc(ruleList []netlink.Rule, src net.IPNet, toCIDRs []string, toFlag bool) error {
+	log.Infof("Update Rule List[%v] for source[%v] with toCIDRs[%v], toFlag[%v]", ruleList, src, toCIDRs, toFlag)
+
+	srcRuleList, err := n.GetRuleListBySrc(ruleList, src)
+
+	if err != nil {
+		log.Errorf("UpdateRuleListBySrc: failed to retrieve rule list %v", err)
+		return err
+	}
+
+	log.Infof("Remove current list [%v]", srcRuleList)
+	var srcRuleTable int
+
+	for _, rule := range srcRuleList {
+		srcRuleTable = rule.Table
+		if n.netLink.RuleDel(&rule); err != nil && !containsNoSuchRule(err) {
+			log.Errorf("Failed to cleanup old IP rule: %v", err)
+			return errors.Wrapf(err, "UpdateRuleListBySrc: failed to delete old rule")
+		}
+
+		var toDst string
+
+		if rule.Dst != nil {
+			toDst = rule.Dst.String()
+		}
+		log.Debugf("UpdateRuleListBySrc: Successfully removed current rule [%v] to %s", rule, toDst)
+	}
+
+	if len(srcRuleList) == 0 {
+		log.Debug("UpdateRuleListBySrc: empty list, no need to update")
+		return nil
+	}
+
+	if toFlag {
```

Comment:
You pass in `!c.networkClient.UseExternalSNAT()` to set this boolean. I think it needs a more descriptive name, maybe something as straight forward as `useExternalSNAT` (and then flip the if-statements of course.)

Reply:
I am trying to keep `useExternalSNAT` on level above UpdateRuleListBySrc() call.